### PR TITLE
Benchmark utilities and first set of benchmark examples

### PR DESCRIPTION
**Pull request by yzhao062**
As discussed, this PR includes a few helper files for benchmarking. The details can be found in benchmark/new_benchmark/README.md. 

Specifically, two helper functions are benchmark_utility.py and profiler_utility.py, which will be used in all benchmarking examples. A corresponding test file is also created (test_benchmark.py).

For now, three sets of benchmarking examples are created: 

- **Node classification** directory contains the code for full graph setting
- **Graph classification** directory contains the code for mini-batch setting
- **Sampler directory** contains the code for neighbor samplers and clusterGCN

Given the current benchmarking code is reasonable, we could extend to more existing PyG examples :)

### CODE REVIEW

File: benchmark/new_benchmark/benchmark_utility.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        return (max(gb_list) * 1024)
+    elif len(mb_list) != 0:
+        return max(mb_list)
+    elif len(kb_list) != 0:
+        return max(kb_list) / 1024
+    else:
+        return max(b_list) / 1024 / 1024
+
+
+def graph_mem_calc(n_nodes, n_edges, node_dims, edge_attr_dims):
```
Comment (mrjel) [inline benchmark/new_benchmark/benchmark_utility.py:219]: Will this work with heterogenous graphs? Is there a possibility that one wants to apply it for heterogenous graphs?

**Review by mrjel — COMMENTED** [review]
- Functionality looks great so far from looking at the code!
- Have some issues running it from command line, as profiler_utility.py is not found by the benchmarking scripts in node/graph-classifcation. Not sure if the folder structure is as it should be (@rusty1s ?).
- What about the empty __init__.py files, should we use them for defining submodules, if not, do we need them?
- Maybe we should integrate the functionality in the benchmark folder, instead of heaving a "new_benchmark"-subfolder.

I am not used to the conventions yet, so I ask you to take all these points (as well as the comments in code) as questions, as my confidence is low right now. :)

File: benchmark/new_benchmark/benchmark_utility.py (hunk 1 above)
Comment (mrjel) [inline benchmark/new_benchmark/benchmark_utility.py:219]: This probably won't work for heterogenous graphs. Do we need it to work with HGs?

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+        # evaluation/inference starts
+        t1 = time.time()
+        # *********************************************************************
+        # make sure the evaluation function is working
+        test_acc = test(model, device, test_loader)
+        # *********************************************************************
+        stats_dict['epoch' + str(epoch)][
+            'test_time'] = time.time() - t1
+
```
[the remark this answers is not in the document]
Comment (yzhao062) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:315]: Great point! I will add synchronization here to see if there is a major difference. or I could switch to torch.cuda.Event + torch.cuda.synchronize() as well :) Will do some experiment off line

File: benchmark/new_benchmark/benchmark_utility.py (hunk 1 above)
Comment (yzhao062) [reply benchmark/new_benchmark/benchmark_utility.py:219]: yeah. this is unlikely to work in heterogeneous settings. We have not covered it in heterogeneous graphs yet. But maybe it is nice to track it as a to-do item.

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rusty1s) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:315]: I think `torch.cuda.Event` is preferred. This allows us to only need to run `torch.cuda.synchronize()` once.

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+# ****************************************************************************
+# Do not change; required for benchmarking
+# Note1: by default we are profiling GPU 0
+# initialize a global profiler
+global_line_profiler = LineProfiler()
+global_line_profiler.enable()
+
+
+class GCNConv(MessagePassing):
```
Comment (rusty1s) [inline benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:49]: We should make use of pre-defined models in order to reduce code duplication, see [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.models.GCN).

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        loss = F.nll_loss(out, batch.y)
+        loss.backward()
+        optimizer.step()
+
+        total_loss += loss.item() * batch.num_graphs
+        total_examples += batch.num_nodes
+    return total_loss / total_examples
+
+
+def train(model, device, train_loader, optimizer):
```
Comment (rusty1s) [inline benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:158]: Do we need an individual script for each model/dataset? What about creating a single `benchmark_graph.py` with customizable models/datasets.

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py (hunk 3 above)
Comment (yzhao062) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:49]: This makes sense. We create customized ones to benchmark back-to-back with other libraries. Some minor things are easier to define in the customized one. I will clean up the duplication when possible since we mainly focus on our performance here.

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py (hunk 4 above)
Comment (yzhao062) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:158]: This is a great point! The only reason we have it in this way is different datasets may need different processes and train/test functions. Happy to update to whatever suits the purpose best.

Comment (rusty1s) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:158]: That‘s true. We can decide which loss function to use based on the utilized dataset.

**Comment by rusty1s** [comment]
Instead of `new_benchmark`, how about we name it `performance` or `efficiency`?

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yzhao062) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:328]: Switched all examples to torch.cuda.Event() instead.

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py (hunk 3 above)
Comment (yzhao062) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:49]: Update the examples with the prebuilt layers when convinient.

File: benchmark/new_benchmark/README.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
+#Benchmark & Profiling
```
[the remark this answers is not in the document]
Comment (yzhao062) [reply benchmark/new_benchmark/README.md:1]: Agreed. I think the whole benchmark directory should be somehow refactored? There is already a benchmark folder. Feel free to share how we could better merge the existing ones and this new one. As @rusty1s has suggested, we could also rename it to performance/efficiency. Not sure what is the most appropriate one. We could defer the renaming as the last step :)

File: benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py (hunk 4 above)
Comment (yzhao062) [reply benchmark/new_benchmark/graph_classification/gcn_enzymes_batch.py:159]: Start adding the unified script to benchmark multiple models for a single dataset. The newly added one (https://github.com/vrtex-team/pytorch_geometric/blob/benchmark/benchmark/kernel/main_performance.py) seems to work multiple models (GCN and SAGE), need to try with all imported models.

File: benchmark/new_benchmark/profiler_utility.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+        "cuda_memory",
+        "occurrences",
+    ],
+)
+
+
+class Profile(object):
+    r"""Layer by layer profiling of PyTorch models, using the PyTorch autograd
+    profiler.
+    It is adapted from torchprof (https://github.com/awwong1/torchprof).
```
Comment (chantat) [inline benchmark/new_benchmark/profiler_utility.py:35]: Saw this in the torchprof github:

```
Attention! This library is deprecated due to the PyTorch 1.9 changes to the torch profiler. Please use the official profiler. Thank you!
```

Probably makes sense to adopt the official pytorch profiler?

File: benchmark/kernel/main_performance.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                    # *****************************************************************
+                    end_train.record()
+                    torch.cuda.synchronize()
+
+                    # recall the first a few epochs may be unstable to measure
+                    # the stats are recorded here
+                    # keep it as it is -> the information is saved in stats_dict
+                    if epoch >= args.warmup_profile:
+                        # store the epoch run information
+                        stats_dict['epoch' + str(epoch)] = {}
```
[the remark this answers is not in the document]
Comment (yzhao062) [reply benchmark/kernel/main_performance.py]: This is a great point! I updated the code in the latest commit (https://github.com/vrtex-team/pytorch_geometric/pull/86/commits/e052ee69d1595a8b6a2f494184e76d57757ec5dd)

File: benchmark/kernel/main_performance.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+                            global_line_profiler)
+
+                        stats_dict['epoch' + str(epoch)][
+                            'max_allocated_cuda'] = curr_mem_usage[0]
+
+                        stats_dict['epoch' + str(epoch)][
+                            'max_reserved_cuda'] = curr_mem_usage[1]
+
+                        stats_dict['epoch' + str(epoch)][
+                            'max_active_cuda'] = curr_mem_usage[2]
```
[the remark this answers is not in the document]
Comment (yzhao062) [reply benchmark/kernel/main_performance.py]: change to namedtuple (https://github.com/vrtex-team/pytorch_geometric/pull/86/commits/e052ee69d1595a8b6a2f494184e76d57757ec5dd). Thanks for the comment.

**Comment by yzhao062** [comment]
Close this PR as discussed with Matthias. Will submit multiple smaller PRs instead. The changes from the comments will carry over.
